File: dynamic-programming/edit-distance.cc
/*
LeetCode 72. Edit Distance

Given two strings word1 and word2, return the minimum number of operations
required to convert word1 to word2.

You have the following three operations permitted on a word:

Insert a character
Delete a character
Replace a character


Example 1:

Input: word1 = "horse", word2 = "ros"
Output: 3
Explanation:
horse -> rorse (replace 'h' with 'r')
rorse -> rose (remove 'r')
rose -> ros (remove 'e')

*/

#include <iostream>
#include <string>
using namespace std;

class Solution {
public:
  int minDistance(string word1, string word2) {

    if (word1.length() == 0)
      return word2.length();
    if (word2.length() == 0)
      return word1.length();

    int dp[word1.length() + 1][word2.length() + 1];

    for (int row = 0; row <= word1.length(); ++row) {
      dp[row][0] = row;
    }
    for (int col = 0; col <= word2.length(); ++col) {
      dp[0][col] = col;
    }
    dp[0][0] = 0;

    for (int row = 1; row <= word1.length(); ++row) {
      for (int col = 1; col <= word2.length(); ++col) {
        dp[row][col] =
            // Insert, deletion, replacement
            std::min(std::min(dp[row][col - 1] + 1, dp[row - 1][col] + 1),
                     dp[row - 1][col - 1] +
                         (word1[row - 1] == word2[col - 1] ? 0 : 1));
      }
    }

    return dp[word1.length()][word2.length()];
  }
};

int main() {
  Solution s;
  cout << s.minDistance("horse", "ros") << endl;
}

File: dynamic-programming/edit-distance.py
'''
LeetCode 72. Edit Distance

Given two strings word1 and word2, return the minimum number of operations
required to convert word1 to word2.

You have the following three operations permitted on a word:

Insert a character
Delete a character
Replace a character


Example 1:

Input: word1 = "horse", word2 = "ros"
Output: 3
Explanation:
horse -> rorse (replace 'h' with 'r')
rorse -> rose (remove 'r')
rose -> ros (remove 'e')
'''

class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        minOps = [[math.inf for _ in range(len(word2)+1)] for _ in range(len(word1)+1)]
        # minOps[i][j] = min ops needed to convert word1[:i] to word2[:j]
        # then minOps[i][j] = min of 
        # - minOps[i][j-1] (add the last char from word2 to end of word1)
        # - minOps[i-1][j] (delete last char from word1)
        # - minOps[i-1][j-1] (replace last char form word1 with last char from word2)
        for i in range(len(word1)+1):
            minOps[i][0] = i
        for j in range(len(word2)+1):
            minOps[0][j] = j

        for i in range(1, len(word1)+1):
            for j in range(1, len(word2)+1):
                if word1[i-1] == word2[j-1]:
                    minOps[i][j] = minOps[i-1][j-1]
                else:
                    minOps[i][j] = min(
                        minOps[i][j-1], 
                        minOps[i-1][j], 
                        minOps[i-1][j-1]
                    ) + 1

        return minOps[len(word1)][len(word2)]

